### PR TITLE
Helm unit tests

### LINKED ISSUE #174: Add Unit Tests for the chart
We have a lot of pain with running topgun/k8s. It would be great to eliminate or greatly reduce it's coverage to strictly integration-level tests only. Most things we want to test can probably be covered by a set of unit tests.

Helm itself does not have a unit testing framework but the community has made a tool to cover this need, [helm-unittest](https://github.com/quintush/helm-unittest)

The original `helm-unittest` project is not maintained (https://github.com/lrills/helm-unittest) but this fork has become the main fork people are contributing to: https://github.com/quintush/helm-unittest

### PR DESCRIPTION
**Pull request by taylorsilva**
# Existing Issue
Fixes #174 

# Why do we need this PR?
We want to stop using topgun/k8s for testing the chart. This starts moving us towards that direction.

# Changes proposed in this pull request
Add a folder called `unit-tests` that uses the `unittest` helm plugin from https://github.com/quintush/helm-unittest. After installing the plugin (run `make install-unittest`) you can run the tests by running `make test` in the root of this repo.

The only unit tests are for the web-deployment and required-check templates. Looking for feedback on the tests written so far and see if this is something we want to keep building/working on.

This also fixes one bug where `web-deployment` was not rendering `strategy` correctly from `values.yam`.

# Contributor Checklist

- [ ] ~Variables are documented in the `README.md`~
- [x] Which branch are you merging into?
    - `master` is for changes related to the current release of the `concourse/concourse:latest` image and should be good to publish immediately
    - `dev` is for changes related to the next release of Concourse (aka unpublished code on `master` in [concourse/concourse](https://github.com/concourse/concourse))

# Reviewer Checklist
> This section is intended for the core maintainers only, to track review progress. Please do not
> fill out this section.
- [ ] Code reviewed
- [ ] Topgun tests run
- [ ] Back-port if needed
- [ ] Is the correct branch targeted? (`master` or `dev`)

### CODE REVIEW

**Review by chenbh — COMMENTED** [review]
This is a pretty neat tool, I can totally see this being a quick and easy sanity check. I'm down for adding this to our PR checks and seeing how it goes.

One thing I don't really like is that the testing is scoped to the file. If a field in values.yml is used in multiple resources, we'll have to write multiple tests. Intuitively, values.yml feels like the "interface" and the unit testing should be based on the "interface" instead of the "internals" like the k8s resources.


Some additional tests we should include:
- `postgresql.enabled = true`
  - a statefulset/service is created
  - the `CONCOURSE_POSTGRES_HOST` is set properly
- tsa related stuff
  - `CONCOURSE_PEER_ADDRESS`
  - `CONCOURSE_TSA_AUTHORIZED_KEYS`
  - `concourse-keys` volumes (this might be hard to test without access to the secret.yml)

**Comment by chenbh** [comment]
Side note: we should probably make use of the `helm test` functionality to run at least the smoke tests when we [lint the chart](https://github.com/concourse/ci/blob/master/tasks/scripts/k8s-lint-and-install)

File: unit-tests/web-deployment-test.yaml [hunk 1]
```diff
@@ -0,0 +1,303 @@
+suite: Web Deployment
+templates:
+  - web-deployment.yaml
+tests:
+  - it: should render successfully with default values.yaml
+    capabilities:
+      # 1.9.0
```
[the remark this answers is not in the document]
Comment (taylorsilva) [reply unit-tests/web-deployment-test.yaml:7]: you learn quite a bit of k8s history when developing a chart :P

**Comment by taylorsilva** [comment]
@chenbh one thing I've been thinking about while writing these tests is: should the tests be very unit-y, as in testing each individual template and conditional in the template OR should they be more behavioural?

For example, for the `fullnameOverride` param I could write this unit test for each template:
```yaml
  - it: fullnameOverride is used
    set:
      fullnameOverride: fullOverride
      web.nameOverride: webOverride
    asserts:
      - equal:
          path: metadata.name
          value: fullOverride-webOverride
```
or I could write one that checks that `fullnameOverride` is used correctly across all templates that it's used in.

Backfilling tests is already hard and I found it easier to write the template-specific unit tests first. I can see us repeating a lot of tests if we continue down this path though. I'm not sure what the best way to structure the tests is then if they're not template specific? Where would the `fullnameOverride` test(s) live? In it's own test suite/file?

**Comment by chenbh** [comment]
> should the tests be very unit-y, as in testing each individual template and conditional in the template OR should they be more behavioural?

I would prefer behaviour, but the unit-y test does have a really nice organizational structure that doesn't pollute my eyes.

I've been thinking about this some more and I'm no longer sure about the need for a lot of the unit tests. A brief look at the commit history shows that the vast majority of PRs are about adding new values to values.yml. Going by the assumption that new fields doesn't really change old template values, the unit tests feel like it'll just be written once, but will never fail cause we never change that part of the template again.

We should have unit tests for large behaviour stuff like worker only deployments, external postgres, etc, but I feel like we _probably_ don't need a unit test for every single env var configuration and other small details

**Comment by taylorsilva** [comment]
> We should have unit tests for large behaviour stuff
I agree with this, it feels too tedious and rigid to continue with this first iteration. When I have some time I'm try writing some tests like that. This was a good first pass to get a feel for what it's like writing these kinds of tests.

**Pull request closed without merging**